tests/Makefile

```diff
@@ -0,0 +1,250 @@
+# Copyright Cartesi and individual authors (see AUTHORS)
+# SPDX-License-Identifier: LGPL-3.0-or-later
+#
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program (see COPYING). If not, see <https://www.gnu.org/licenses/>.
+#
+
+PREFIX = /opt/cartesi
+INSTALLDIR = $(PREFIX)/tests
+
+LUA_BIN ?= $(shell which lua5.4)
+
+TOOLCHAIN_DOCKER_REPOSITORY ?= cartesi/toolchain
+TOOLCHAIN_TAG ?= 0.16.0
+
+BUILDDIR = $(abspath build)
+RISCV_PREFIX = riscv64-cartesi-linux-gnu-
+
+MACHINE_EMULATOR_SRC_DIR = $(abspath ../src)
+
+MACHINE_EMULATOR_SO_VERSION= $(shell make -sC ../src so-version)
+
+TARGETS_WITH_TOOLCHAIN = $(addsuffix -with-toolchain,build-tests-machine build-tests-uarch machine uarch misc riscv-tests riscv-arch-test)
+
+CARTESI_IMAGES_PATH ?= $(abspath ./build/images)
+CARTESI_TESTS_PATH ?= $(abspath ./build/machine)
+CARTESI_TESTS_UARCH_PATH ?= $(abspath ./build/uarch)
+
+export CARTESI_IMAGES_PATH
+export CARTESI_TESTS_PATH
+export CARTESI_TESTS_UARCH_PATH
+
+TARGET_OS ?= $(shell uname -s)
+ifeq ($(TARGET_OS),Darwin)
+LIBCARTESI_SO=libcartesi-$(MACHINE_EMULATOR_SO_VERSION).dylib
+LIBCARTESI_SO_GRPC=libcartesi_grpc-$(MACHINE_EMULATOR_SO_VERSION).dylib
+LIBCARTESI_SO_JSONRPC=libcartesi_jsonrpc-$(MACHINE_EMULATOR_SO_VERSION).dylib
+
+NUM_JOBS := $(shell sysctl -n hw.ncpu)
+
+LIBASAN_SO=$(shell realpath `$(CC) -print-file-name=libclang_rt.asan_osx_dynamic.dylib`)
+LD_PRELOAD_PREFIX=DYLD_INSERT_LIBRARIES="$(LIBASAN_SO)"
+else
+LIBCARTESI_SO=libcartesi-$(MACHINE_EMULATOR_SO_VERSION).so
+LIBCARTESI_SO_GRPC=libcartesi_grpc-$(MACHINE_EMULATOR_SO_VERSION).so
+LIBCARTESI_SO_JSONRPC=libcartesi_jsonrpc-$(MACHINE_EMULATOR_SO_VERSION).so
+
+NUM_JOBS := $(shell nproc)
+
+LIBASAN_SO=$(shell realpath `$(CC) -print-file-name=libasan.so`)
+LIBSTDCPP_SO=$(shell realpath `$(CC) -print-file-name=libstdc++.so`)
+LD_PRELOAD="$(LIBASAN_SO) $(LIBSTDCPP_SO)"
+LD_PRELOAD_PREFIX=LD_PRELOAD=$(LD_PRELOAD)
```

Comment:
I had crash while testing, this line is always preloading the sanitizer shared library, even though I did not compile and was running with sanitizer enabled. And giving me errors because things were compiled without sanitizer:

```
2024-01-31 12-52-40 info remote-cartesi-machine pid:1232149 ppid:1226943 Initializing server on 127.0.0.1:7001

*** No errors detected
LD_PRELOAD="/usr/lib/libasan.so.8.0.0 /usr/lib/libstdc++.so.6.0.32" ./build/misc/test-merkle-tree-hash --log2-root-size=30 --log2-leaf-size=12 --input=build/misc/test-merkle-tree-hash
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1232222==ERROR: AddressSanitizer: SEGV on unknown address 0x63ff9e3a3e48 (pc 0x7b697885738f bp 0x000000000000 sp 0x7fff21f26100 T0)
==1232222==The signal is caused by a READ memory access.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
make[1]: *** [Makefile:187: test-hash] Error 1
make[1]: Leaving directory '/home/bart/projects/cartesi/machine-emulator-sdk/emulator/tests'
make: *** [Makefile:243: test] Error 2
```

Reply:
thanks